compiler/record-minmax/src/MinMaxObserver.cpp

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2020 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "MinMaxObserver.h"
+
+#include <luci/IR/CircleOpcode.h>
+
+using DataType = luci_interpreter::DataType;
+
+namespace record_minmax
+{
+
+// postTensorWrite is only called for a node producing a tensor
+void MinMaxObserver::postTensorWrite(const luci::CircleNode *node,
+                                     const luci_interpreter::Tensor *tensor)
+{
+  // CircleOutput does not produce a tensor
+  assert(node->opcode() != luci::CircleOpcode::CIRCLEOUTPUT);
+
+  // Operators with multiple outputs
+  assert(node->opcode() != luci::CircleOpcode::IF);
+  assert(node->opcode() != luci::CircleOpcode::SPLIT);
+  assert(node->opcode() != luci::CircleOpcode::SPLIT_V);
+  assert(node->opcode() != luci::CircleOpcode::TOPK_V2);
+  assert(node->opcode() != luci::CircleOpcode::UNPACK);
+  assert(node->opcode() != luci::CircleOpcode::WHILE);
+
+  if (node->opcode() == luci::CircleOpcode::CONST)
+  {
+    // node is not activation. Do nothing.
+    return;
+  }
+
+  if (node->opcode() == luci::CircleOpcode::ARG_MAX)
+  {
+    // Output of arg_max is the index of the largest value across axes of a tensor
+    // this should not be quantized
+    return;
+  }
+
+  // Only support recording of float32 values
+  if (tensor->element_type() != DataType::FLOAT32)
```

Comment:
I'm not sure if we need to throw. I think it is legal to have some ops working with floats and others working with other types. Those working with different types should just be skipped.

Reply:
A non-float model is a wrong input to `record-minmax`. I don't want to allow users to run `record-minmax` with a non-float model.

Reply:
Float model can contain integer tensors.
For example, Transpose has input that specifies the order of axis. There are many such operators. Are you going to blacklist all of them?

Reply:
Sorry for the confusion. I used the wrong term (float model). The input expected by `record-minmax` is a model whose "activation" tensors are float.
(Activation indicates the output of an operator; the word 'activation' originated from the output of fully connected layer (which has an activation function), but it seems to be used for other operators now.)

The purpose of `record-minmax` is to record the min/max of the activation tensors for quantization. In my thought, a model with non-float activation tensors (except some special Ops such as ArgMax) can be seen as a quantized model (either partially or fully). So I throw an exception, because already quantized model is definitely the wrong input to the `record-minmax`).

The axis of Transpose is not the activation tensor, because it is not the output of any operator but just a constant tensor (CircleConst). So, it is ok that the axis (non-activation) is an integer-typed tensor. And we do not record the values of the axis, because it is not the activation.
(It is filtered by the [earlier if statement](https://github.com/Samsung/ONE/pull/1867/files#diff-c9aac1a089f2141b49ed7ae04af7c774R41))

I think it is necessary to maintain the blacklist of Ops whose output tensor (activation) cannot be quantized. We will not record the min/max values of those Ops. For now, the blacklist only includes ArgMax, but I think there are more (e.g., ArgMin, etc. do you have any idea on such Ops?). I will add the guard code for them later.

Reply:
Thank you for detailed explanation.
> The axis of Transpose is not the activation tensor, because it is not the output of any operator but just a constant tensor (CircleConst).

It is not necessarily constant, it can be output of another operator, such as Concatenation (I've seen model with it). But we cannot blacklist Concatenation, because it can also work with "activations".
Let's continue the discussion elsewhere